File: crates/superposition_provider/examples/local_with_fallback_example.rs
use std::path::PathBuf;

use open_feature::{EvaluationContext, OpenFeature};
use superposition_provider::{
    data_source::file::FileDataSource, data_source::http::HttpDataSource,
    local_provider::LocalResolutionProvider, PollingStrategy, RefreshStrategy,
    SuperpositionOptions,
};
use tokio::time::{sleep, Duration};

#[tokio::main]
async fn main() {
    env_logger::init();

    let http_source = HttpDataSource::new(SuperpositionOptions::new(
        "http://localhost:8080".to_string(),
        "token".to_string(),
        "localorg".to_string(),
        "dev".to_string(),
    ));

    let manifest_dir = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    let file_source = FileDataSource::new(manifest_dir.join("examples/config.toml"));

    let provider = LocalResolutionProvider::new(
        Box::new(http_source),
        Some(Box::new(file_source)),
        RefreshStrategy::Polling(PollingStrategy {
            interval: 10,
            timeout: Some(10),
        }),
    );

    // Register with OpenFeature and create a client
    let mut api = OpenFeature::singleton_mut().await;
    api.set_provider(provider).await;
    let client = api.create_client();

    // Allow time for the provider to initialize via OpenFeature
    sleep(Duration::from_secs(2)).await;

    println!("=== Superposition Fallback + Polling Example ===");
    println!("Primary: HTTP (localhost:8080), Fallback: config.toml");
    println!("Polling every 10s. Printing config every 5s (Ctrl-C to stop).\n");

    let context = EvaluationContext::default()
        .with_targeting_key("user-456")
        .with_custom_field("os", "linux")
        .with_custom_field("city", "Berlin");

    loop {
        let ts = chrono::Utc::now().format("%H:%M:%S");

        match client
            .get_string_value("currency", Some(&context), None)
            .await
        {
            Ok(value) => print!("[{}] currency = {}", ts, value),
            Err(e) => print!("[{}] currency error: {:?}", ts, e),
        }

        match client.get_int_value("timeout", Some(&context), None).await {
            Ok(value) => println!("  |  timeout = {}", value),
            Err(e) => println!("  |  timeout error: {:?}", e),
        }

        sleep(Duration::from_secs(5)).await;
    }
}

File: clients/python/provider/examples/local_with_fallback_example.py
"""
Local with Fallback Example - Use HTTP with local file as fallback.

Demonstrates resilience with primary + fallback data sources.
If the HTTP server fails, the provider automatically falls back to local file config.

Prerequisites:
  - Superposition server (optional) at http://localhost:8080
  - config.toml file for fallback
"""

import asyncio
import logging
from pathlib import Path

from openfeature.evaluation_context import EvaluationContext

from superposition_provider import LocalResolutionProvider, HttpDataSource, FileDataSource
from superposition_provider.types import SuperpositionOptions, PollingStrategy

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


async def main():
    """Run the local with fallback example."""
    # Get the path to the example config file
    example_dir = Path(__file__).parent
    config_path = example_dir / "config.toml"

    logger.info("=== Superposition Fallback + Polling Example ===")
    logger.info(f"Primary: HTTP (localhost:8080)")
    logger.info(f"Fallback: {config_path}")
    logger.info("Polling every 10s.\n")

    # Configure Superposition API connection
    options = SuperpositionOptions(
        endpoint="http://localhost:8080",
        token="token",
        org_id="localorg",
        workspace_id="dev",
    )

    # Create a provider with both HTTP (primary) and file (fallback) sources
    provider = LocalResolutionProvider(
        primary_source=HttpDataSource(options),
        refresh_strategy=PollingStrategy(interval=10, timeout=10),
        fallback_source=FileDataSource(config_path.__str__()),
    )

    # Initialize the provider
    await provider.initialize(EvaluationContext())

    # Create an evaluation context
    context = EvaluationContext(
        targeting_key="user-456",
        attributes={
            "os": "linux",
            "city": "Berlin",
        },
    )

    logger.info("Polling config every 5s. Press Ctrl-C to stop.\n")

    try:
        # Poll in a loop to show resolved config
        while True:
            logger.info("Fetching config...")
            config = provider.resolve_all_features(context)
            logger.info(f"Resolved config: {config}")

            # Also show a specific value
            currency = provider.resolve_string_details("currency", "No value", context)
            logger.info(f"Currency: {currency.value}\n")

            await asyncio.sleep(5)
    except KeyboardInterrupt:
        logger.info("\nShutting down...")
    finally:
        await provider.shutdown()
        logger.info("Provider shutdown complete")


if __name__ == "__main__":
    asyncio.run(main())
